src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -174,27 +174,31 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
         var targetExpression = Expression.Convert(TargetExpr, methodInfo.DeclaringType);
         var targetableRequestDelegate = CreateTargetableRequestDelegate(methodInfo, targetExpression, factoryContext, context => targetFactory(context));
 
-        return new RequestDelegateResult(httpContext => targetableRequestDelegate(targetFactory(httpContext), httpContext), factoryContext.Metadata);
+        return new RequestDelegateResult(httpContext => targetableRequestDelegate(targetFactory(httpContext), httpContext), GetMetadataList(factoryContext.Metadata));
     }
 
     private static FactoryContext CreateFactoryContext(RequestDelegateFactoryOptions? options)
     {
-        var context = new FactoryContext
+        return new FactoryContext
         {
             ServiceProvider = options?.ServiceProvider,
             ServiceProviderIsService = options?.ServiceProvider?.GetService<IServiceProviderIsService>(),
             RouteParameters = options?.RouteParameterNames?.ToList(),
             ThrowOnBadRequest = options?.ThrowOnBadRequest ?? false,
             DisableInferredFromBody = options?.DisableInferBodyFromParameters ?? false,
-            Filters = options?.RouteHandlerFilterFactories?.ToList()
+            Filters = options?.RouteHandlerFilterFactories?.ToList(),
+            Metadata = options?.EndpointMetadata ?? new List<object>(),
         };
+    }
 
-        if (options?.InitialEndpointMetadata is not null)
+    private static List<object> GetMetadataList(IList<object> metadata)
```

Comment:
I'm a little confused by this method. Isn't `metadata` going to be a `List<object>` already if we're reading it from the factory context?

Reply:
Nevermind. I see the `FactoryContext.Metadata` type has been changed to an `IList`. Why not use the concrete type there?

Reply:
It's because we're taking in a mutable `IList<object>` which has to match `EndpointBuilder.Metadata` and we need to return an `IReadOnlyList<object>` which is required by the already-shipped `RequestDelegateResult.RequestDelegateResult`. If it's just a `List<object>` like it is most of the time we can cast it and treat it as an `IReadOnlyList<object>`.

I also could have just tried `metadata is IReadOnlyList<object> readOnlyList` and returned that if it succeeded and only return `new List<object>(metadata)` as a last resort. Not sure why I didn't do it that way to begin with. Fixed. I'm not calling it `AsReadOnlyList` instead to make the intent more clear.